### PR TITLE
Can

### PR DESCRIPTION
**Pull request by mjkpolo**
Preliminary design for c++ified old code

### CODE REVIEW

File: embedded/drivers/include/can.h
Comment (bennyE31) [inline embedded/drivers/include/can.h]: Let's drop this, since it isn't even used in pod-embedded. We can add it back if we ever feel like we need it

File: embedded/drivers/include/can.h
Comment (bennyE31) [inline embedded/drivers/include/can.h]: Pedantic, but let's try to keep includes that aren't required by the header file in the source file. Also drop the time.h include along with the timer variable

**Review by bennyE31 — CHANGES_REQUESTED** [review]
Looks good functionally. Let's try to get some more comments before we merge. Like Eric said, they were rushed when they wrote this, but we have more time to document our changes than they did. I think block comments in the headers describing the classes are a good idea, as well as maybe some inline comments on the class members. Function headers are usually good, but  idk functions like `canSend` and `canRead` are pretty self explanatory. 

Also, we had discussed dropping the term "RMS", and I think this is a good spot to do that. However, we might need approval beyond @EUdds for that.

File: embedded/drivers/include/can.h
Comment (bennyE31) [reply embedded/drivers/include/can.h]: by "this" I mean the `itimerval`

Comment (mjkpolo) [reply embedded/drivers/include/can.h]: EDIT:

Lmao I somehow missed your second comment. But that was a bit confusing haha

File: embedded/drivers/include/can.h
Comment (mjkpolo) [reply embedded/drivers/include/can.h]: Ok, dropped the timer var and time header file.

You're sure you don't want the headers to have all the includes so the source is cleaner and just says include "can.h"? I thought new members would be less intimidated and understand better having all defines and header files in the .h. If you notice rms.h is cleaner in my opinion too having the includes in can.h
I personally also think it's easier to keep track of files we add and remove to have them concentrated in an upper most header basically.

What do you think @EUdds

File: embedded/devices/README.md
Comment (EUdds) [inline embedded/devices/README.md]: As we noted, this is our chance to drop the RMS acronym for good, lets discuss this on slack.

File: embedded/devices/include/rms.h
Comment (EUdds) [inline embedded/devices/include/rms.h]: It seems kinda silly to define 3 different macros that all mean make the motor go forward at 0,10,20,30 NM.

The datasheet defines this as the COMMAND_MESSAGE id. Lets call it that instead

File: embedded/devices/src/rms.cpp
Comment (EUdds) [inline embedded/devices/src/rms.cpp]: Will timing information come a little later?

File: embedded/devices/src/rms.cpp
Comment (EUdds) [inline embedded/devices/src/rms.cpp]: I'm not a huge fan of segments of abstract commented out code here. I understand the desire to keep it. I'm torn about this.

Could you at least add comments on what each ID carries in it's data bytes and we'll go from there as the season start?

File: embedded/drivers/README.md
Comment (EUdds) [inline embedded/drivers/README.md]: Can you give a bit more detail on how the queue system works and why there is nothing a developer needs to do on the CAN driver side of things

**Review by EUdds — CHANGES_REQUESTED** [review]
Overall I think this is a great start. I really think though this is our time to finally kill the RMS acronym. I'm going to tie in @lmaddox28 @dawiggleman and @keving98g and we can discuss.

**Comment by EUdds** [comment]
Also you might not have pre-commit configured correctly if you're failing lint. Run `./bloop-dev lint`

File: embedded/devices/src/rms.cpp
Comment (mjkpolo) [reply embedded/devices/src/rms.cpp]: 😳 plagiarism exposed ‼️

File: embedded/drivers/include/can.h
Comment (bennyE31) [reply embedded/drivers/include/can.h]: From [GCC](https://gcc.gnu.org/onlinedocs/cpp/Header-Files.html#:~:text=A%20header%20file%20is%20a,Header%20files%20serve%20two%20purposes)
"A header file is a file containing C declarations and macro definitions to be shared between several source files."

Headers are meant to be shared between source files. The includes that are specific to the source file might not need to be included in other sources that include the header file

**Comment by bennyE31** [comment]
Do we want to push the RMS specific changes into the repository when we're still unsure of our motor controller choice for the upcoming competition? 

The framework is good, but what would you guys think of  replacing the RMS specific code with dummy functions so things still build and we have a skeleton for whatever motor controller we end up using?

I don't want anyone to lose any work, of course, so I think a good plan would be to include those changes in a separate pull request we can leave open until we make a decision, and merge the CAN drivers and MotorController skeleton into main. 

Then, in the future, if we pick a different motor controller, we just have to fill out a skeleton, or, if we stick with the RMS controller, we can just click merge on the open pull request.

This would enforce the idea that the main branch is meant to be our final product.

Thoughts? @EUdds @mjkpolo

**Comment by EUdds** [comment]
> 
> 
> Do we want to push the RMS specific changes into the repository when we're still unsure of our motor controller choice for the upcoming competition?
> 
> The framework is good, but what would you guys think of replacing the RMS specific code with dummy functions so things still build and we have a skeleton for whatever motor controller we end up using?
> 
> I don't want anyone to lose any work, of course, so I think a good plan would be to include those changes in a separate pull request we can leave open until we make a decision, and merge the CAN drivers and MotorController skeleton into main.
> 
> Then, in the future, if we pick a different motor controller, we just have to fill out a skeleton, or, if we stick with the RMS controller, we can just click merge on the open pull request.
> 
> This would enforce the idea that the main branch is meant to be our final product.
> 
> Thoughts? @EUdds @mjkpolo

I think since we're making CAN such that adding / removing devices from the driver is as simple as instanciating a class. There is no harm in keeping the driver here. We aren't really cutting it close on repo space and if the file isn't used somewhere it doesn't make the executable larger. This has a very high chance of just being the template for adding CAN devices, but I think that having the code in there as it would be if it were an actual used device would help contributing to similar CAN devices.

In the end it's your call.

File: embedded/devices/src/motorController.cpp
Comment (EUdds) [inline embedded/devices/src/motorController.cpp]:
```suggestion
  uint8_t payload[] = {TORQUE_SCALE_LWR(20), 0x0, 0x0, 0x0, 0x1, 0x1, 0x0, 0x0};
```

File: embedded/devices/src/motorController.cpp
Comment (EUdds) [inline embedded/devices/src/motorController.cpp]:
```suggestion
  uint8_t payload[] = {TORQUE_SCALE_LWR(30), 0x0, 0x0, 0x0, 0x1, 0x1, 0x0, 0x0};
```

**Review by EUdds — CHANGES_REQUESTED** [review]

**Comment by bennyE31** [comment]
Alright, let's just keep everything

**Comment by bennyE31** [comment]
Closing this PR.

We have a much simpler CAN bus than we anticipated last summer. This ended up bringing along a simpler solution

**Pull request closed without merging**
